File: frontend/src/services/transaction.service.ts
import { api } from './api';
import { ApiResponse } from '@/types/auth.types';
import { DashboardSummary, PageResponse, Transaction, TransactionRequest } from '@/types/transaction.types';

export const transactionService = {
  async getDashboardSummary(): Promise<DashboardSummary> {
    const res = await api.get<ApiResponse<DashboardSummary>>('/user/dashboard/summary');
    return res.data.data;
  },

  async getTransactions(params?: {
    page?: number;
    size?: number;
    type?: string;
    categoryId?: string;
    walletId?: string;
    startDate?: string;
    endDate?: string;
  }): Promise<PageResponse<Transaction>> {
    const res = await api.get<ApiResponse<PageResponse<Transaction>>>('/user/transactions', { params });
    return res.data.data;
  },

  async getTransactionById(id: string): Promise<Transaction> {
    const res = await api.get<ApiResponse<Transaction>>(`/user/transactions/${id}`);
    return res.data.data;
  },

  async createTransaction(data: TransactionRequest): Promise<Transaction> {
    const res = await api.post<ApiResponse<Transaction>>('/user/transactions', data);
    return res.data.data;
  },

  async createTransactionsBatch(data: TransactionRequest[]): Promise<Transaction[]> {
    const res = await api.post<ApiResponse<Transaction[]>>('/user/transactions/batch', data);
    return res.data.data;
  },

  async updateTransaction(id: string, data: TransactionRequest): Promise<Transaction> {
    const res = await api.put<ApiResponse<Transaction>>(`/user/transactions/${id}`, data);
    return res.data.data;
  },

  async deleteTransaction(id: string): Promise<void> {
    await api.delete(`/user/transactions/${id}`);
  },
};

File: backend/src/main/java/com/financialrecord/service/TransactionService.java
package com.financialrecord.service;

import com.financialrecord.dto.request.TransactionRequest;
import com.financialrecord.dto.response.DashboardSummaryResponse;
import com.financialrecord.dto.response.TransactionResponse;
import com.financialrecord.entity.enums.TransactionType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.UUID;

public interface TransactionService {

    Page<TransactionResponse> getTransactions(
            UUID userId,
            TransactionType type,
            UUID categoryId,
            UUID walletId,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable
    );

    TransactionResponse getTransactionById(UUID id, UUID userId);

    TransactionResponse createTransaction(UUID userId, TransactionRequest request);

    java.util.List<TransactionResponse> createTransactionsBatch(UUID userId, java.util.List<TransactionRequest> requests);

    TransactionResponse updateTransaction(UUID id, UUID userId, TransactionRequest request);

    void deleteTransaction(UUID id, UUID userId);

    DashboardSummaryResponse getDashboardSummary(UUID userId);
}
